### ISSUE TITLE
(CloudInit) Static IP address not getting applied until VM is manually rebooted

### ISSUE THREAD

**Issue by davidedpg10** [Telmate/terraform-provider-proxmox#603]
Hello, I'm having an issue where the IP Address does not get applied until I manually go reboot the VM. The settings get set correctly in CloudInit, it just doesn't pick it up right away. I can't even use `remote-exec` provisioner to reboot it because originally the VMs pick up a DHCP address (making me unable to predict the connection values), once rebooted they set the static IP to the value specified in CloudInit. Is this a bug? expected behavior?

This is what the resource looks like in the template
```
resource "proxmox_vm_qemu" "kube-agent" {
  count = 2
  name = "kube-agent-0${count.index + 1}"
  target_node = "pve"
  vmid = "801${count.index + 1}"
  clone = "ubuntu-server-jammy-template-base"
  agent = 1
  os_type = "cloud-init"
  cores = 2
  sockets = 1
  cpu = "host"
  memory = 4096
  scsihw = "virtio-scsi-pci"
  bootdisk = "scsi0"
  disk {
    slot = 0
    size = "20G"
    type = "scsi"
    storage = "local-lvm"
    #storage_type = "zfspool"
    iothread = 1
  }
  network {
    model = "virtio"
    bridge = "vmbr0"
  }

  network {
    model = "virtio"
    bridge = "vmbr17"
  }

  ipconfig0 = "ip=10.0.0.21${count.index + 1}/24,gw=10.0.0.254"
  ipconfig1 = "ip=10.17.0.5${count.index + 1}/24"
  sshkeys = <<EOF
  ${var.ssh_pub_key}
  EOF
}
```

**Comment by stefangrosaru** [comment]
Experiencing the same issue

**Comment by mleone87** [comment]
Is the template customized or is a plain cloudinit image?

**Comment by stefangrosaru** [comment]
@mleone87 in my case the template used is built using packer from a ubuntu 22 server image.
@davidedpg10 I was able to use `self.ssh_host` in `remote-exec` in order to reboot the vm after it's built, I think you need to make sure that qemu-guest-agent is running on the vm so that it can report its ip.

**Comment by RaceFPV** [comment]
I ran into the same bug, but watch out because even after forcing a reboot after creation, any hardware change via terraform that causes the vm to reboot will make it grab a new ip addy

**Comment by deespater** [comment]
Issue still persists on 2.9.13

**Comment by bubthegreat** [comment]
To add to this problem, there's a race condition in provisioning so the DHCP assigment is assigning multiple VMs to the same IP address, preventing the remote-exec with a connection from connecting to the VM to execute a reboot and force the application of the updated IP addresses.

I'm working around this currently by using a null resource remote-exec to the proxmox server that will reboot the VM through the proxmox command line, but this should really be done through the API properly - this was just an easy enough workaround.
```

resource "null_resource" "k8s_server_1_reboot" {
  depends_on    = [proxmox_vm_qemu.k8s_server_1]
  provisioner "remote-exec" {
    inline = [
      "qm reboot ${proxmox_vm_qemu.k8s_server_1.vmid}",
    ]
    connection {
      type     = "ssh"
      user     = "${var.proxmox_user}"
      password = "${var.proxmox_password}"
      host     = "${var.proxmox_ip}"
      timeout  = "60s"
    }
  }
}
```

**Comment by zimmertr** [comment]
Ditto with the Debian Openstack image.

**Comment by bubthegreat** [comment]
This should not be marked as stale, it's still an issue.

**Comment by dylanbegin** [comment]
Adding `ci_wait = 40` fixed it for me on Rocky Linux 9.2. I think cloud-init is starting before the qemu-guest-agent service is running.  

Also, make sure your packer config has the lines below in the shell provisioner block to ensure the qemu service is set to start on boot:
` "sudo systemctl start qemu-guest-agent",`
` "sudo systemctl enable qemu-guest-agent",`

**Comment by xlanor** [comment]
Still an issue on ubuntu24 and latest rc proxmox provider.
